source/common/http/http2/codec_impl.cc

```diff
@@ -349,6 +354,36 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
   }
 }
 
+void ConnectionImpl::StreamImpl::processBufferedData(bool stream_being_destroyed) {
+  ENVOY_CONN_LOG(debug, "Stream {} processing buffered data.", parent_.connection_, stream_id_);
+
+  auto next_stage = stream_manager_.getNextStage();
+  stream_manager_.flush_all_data_ = stream_being_destroyed;
+  bool continue_processing = stream_being_destroyed || !buffersOverrun();
+  while (continue_processing && next_stage.has_value()) {
+    switch (next_stage.value()) {
+    case BackedUpStreamManager::Stage::Body:
+      if (stream_manager_.data_end_stream_) {
+        remote_end_stream_ = true; // restore the deferred end_stream from the remote.
+      } else if (stream_manager_.trailers_buffered_) {
+        // Avoid inversion in the case where we saw trailers, acquiring the
+        // remote_end_stream_ being set to true, but the trailers ended up being
+        // buffered. The buffered decodeTrailers call will restore the
+        // remote_end_stream_.
+        remote_end_stream_ = false;
```

Comment:
I don't love overloading remote_end_stream.  Can we put this logic in decodeData as send_end_stream = got_end_stream + no_trailers?